In [1]:
import pandas as pd

postings = pd.read_csv("../data/raw/archive/postings.csv", low_memory=False)

In [2]:
postings.head()

,job_id,company_name,title,description,max_salary,pay_period,location,company_id,views,med_salary,...,skills_desc,listed_time,posting_domain,sponsored,work_type,currency,compensation_type,normalized_salary,zip_code,fips
0,921716,Corcoran Sawyer Smith,Marketing Coordinator,Job descriptionA leading real estate firm in N...,20.0,HOURLY,"Princeton, NJ",2774458.0,20.0,NaN,...,Requirements: \n\nWe are seeking a College or ...,1.713398e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,38480.0,8540.0,34021.0
1,1829192,NaN,Mental Health Therapist/Counselor,"At Aspen Therapy and Wellness , we are committ...",50.0,HOURLY,"Fort Collins, CO",NaN,1.0,NaN,...,NaN,1.712858e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,83200.0,80521.0,8069.0
2,10998357,The National Exemplar,Assitant Restaurant Manager,The National Exemplar is accepting application...,65000.0,YEARLY,"Cincinnati, OH",64896719.0,8.0,NaN,...,We are currently accepting resumes for FOH - A...,1.713278e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,55000.0,45202.0,39061.0
3,23221523,"Abrams Fensterman, LLP",Senior Elder Law / Trusts and Estates Associat...,Senior Associate Attorney - Elder Law / Trusts...,175000.0,YEARLY,"New Hyde Park, NY",766262.0,16.0,NaN,...,This position requires a baseline understandin...,1.712896e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,157500.0,11040.0,36059.0
4,35982263,NaN,Service Technician,Looking for HVAC service tech with experience ...,80000.0,YEARLY,"Burlington, IA",NaN,3.0,NaN,...,NaN,1.713452e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,70000.0,52601.0,19057.0


In [3]:
postings.shape

(123849, 31)

In [4]:
postings.columns.tolist()

['job_id',
 'company_name',
 'title',
 'description',
 'max_salary',
 'pay_period',
 'location',
 'company_id',
 'views',
 'med_salary',
 'min_salary',
 'formatted_work_type',
 'applies',
 'original_listed_time',
 'remote_allowed',
 'job_posting_url',
 'application_url',
 'application_type',
 'expiry',
 'closed_time',
 'formatted_experience_level',
 'skills_desc',
 'listed_time',
 'posting_domain',
 'sponsored',
 'work_type',
 'currency',
 'compensation_type',
 'normalized_salary',
 'zip_code',
 'fips']

In [5]:
postings.info()

<class 'pandas.DataFrame'>
RangeIndex: 123849 entries, 0 to 123848
Data columns (total 31 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   job_id                      123849 non-null  int64  
 1   company_name                122130 non-null  str    
 2   title                       123849 non-null  str    
 3   description                 123842 non-null  str    
 4   max_salary                  29793 non-null   float64
 5   pay_period                  36073 non-null   str    
 6   location                    123849 non-null  str    
 7   company_id                  122132 non-null  float64
 8   views                       122160 non-null  float64
 9   med_salary                  6280 non-null    float64
 10  min_salary                  29793 non-null   float64
 11  formatted_work_type         123849 non-null  str    
 12  applies                     23320 non-null   float64
 13  original_listed_time     

In [6]:
postings.describe()

,job_id,max_salary,company_id,views,med_salary,min_salary,applies,original_listed_time,remote_allowed,expiry,closed_time,listed_time,sponsored,normalized_salary,zip_code,fips
count,1.238490e+05,2.979300e+04,1.221320e+05,122160.000000,6280.000000,2.979300e+04,23320.000000,1.238490e+05,15246.0,1.238490e+05,1.073000e+03,1.238490e+05,123849.0,3.607300e+04,102977.000000,96434.000000
mean,3.896402e+09,9.193942e+04,1.220401e+07,14.618247,22015.619876,6.491085e+04,10.591981,1.713152e+12,1.0,1.716213e+12,1.712928e+12,1.713204e+12,0.0,2.053270e+05,50400.491887,28713.879887
std,8.404355e+07,7.011101e+05,2.554143e+07,85.903598,52255.873846,4.959738e+05,29.047395,4.848209e+08,0.0,2.321394e+09,3.622893e+08,3.989122e+08,0.0,5.097627e+06,30252.232515,16015.929825
min,9.217160e+05,1.000000e+00,1.009000e+03,1.000000,0.000000,1.000000e+00,1.000000,1.701811e+12,1.0,1.712903e+12,1.712346e+12,1.711317e+12,0.0,0.000000e+00,1001.000000,1003.000000
25%,3.894587e+09,4.828000e+01,1.435200e+04,3.000000,18.940000,3.700000e+01,1.000000,1.712863e+12,1.0,1.715481e+12,1.712670e+12,1.712886e+12,0.0,5.200000e+04,24112.000000,13121.000000
50%,3.901998e+09,8.000000e+04,2.269650e+05,4.000000,25.500000,6.000000e+04,3.000000,1.713395e+12,1.0,1.716042e+12,1.712670e+12,1.713408e+12,0.0,8.150000e+04,48059.000000,29183.000000
75%,3.904707e+09,1.400000e+05,8.047188e+06,8.000000,2510.500000,1.000000e+05,8.000000,1.713478e+12,1.0,1.716088e+12,1.713283e+12,1.713484e+12,0.0,1.250000e+05,78201.000000,42077.000000
max,3.906267e+09,1.200000e+08,1.034730e+08,9975.000000,750000.000000,8.500000e+07,967.000000,1.713573e+12,1.0,1.729125e+12,1.713562e+12,1.713573e+12,0.0,5.356000e+08,99901.000000,56045.000000


In [7]:
postings["company_name"].nunique()

24428

In [8]:
postings["location"].value_counts().head(10)

location
United States    8125
New York, NY     2756
Chicago, IL      1834
Houston, TX      1762
Dallas, TX       1383
Atlanta, GA      1363
Boston, MA       1176
Austin, TX       1083
Charlotte, NC    1075
Phoenix, AZ      1059
Name: count, dtype: int64

In [9]:
postings["title"].value_counts().head(20)

title
Sales Manager                      673
Customer Service Representative    373
Project Manager                    354
Administrative Assistant           254
Senior Accountant                  238
Executive Assistant                228
Salesperson                        211
Registered Nurse                   210
Receptionist                       204
Staff Accountant                   200
Account Executive                  195
Retail Sales Associate             190
Sales Associate                    189
Software Engineer                  181
Controller                         175
Account Manager                    171
Store Manager                      166
Senior Software Engineer           162
Assistant Manager                  161
ASSISTANT STORE MANAGER            161
Name: count, dtype: int64

In [10]:
postings["max_salary"].notna().sum() / len(postings) * 100

np.float64(24.05590678971974)

## Initial EDA Observations — 2026-10-02

### Basic facts
- postings: 123,849 rows × 31 columns
- job_skills: 213,768 rows (job_id → skill_abr mapping)
- skills: 35 rows (skill_abr → skill category name)
- postings_with_skills (after merge): 207,531 rows × 33 columns

### Salary data
- Salary present in only 24.06% of postings (~76% MISSING)
- Salary outliers observed:
  - max_salary max = 120,000,000 (unrealistic)
  - min_salary min = 1 (unrealistic)
  - normalized_salary max = 535,600,000 (unrealistic)
- Cleanup required in Phase 3 (SRS Section 4.4)

### Locations & titles
- Unique companies: 24,428
- Top locations: United States (8125), New York NY (2756), Chicago IL (1834)
- Top job titles: Sales Manager (673), Customer Service Rep (373), Project Manager (354)

### ⚠️ IMPORTANT — skills in this dataset are CATEGORIES, not specific skills
- skills.csv contains 35 BROAD categories:
  Information Technology, Sales, Management, Manufacturing,
  Health Care Provider, Business Development, Engineering, Other,
  Finance, Marketing, Accounting/Auditing, Administrative, ...
- These are NOT specific skills like Python, SQL, JavaScript.
- For JMIS's specific-skill analysis (Skill Demand Score, resume gaps,
  job recommendations), we will extract specific skills from the
  job `description` column in Phase 5 using a custom skills dictionary.
- Both levels will be used:
  * Coarse categories → industry/department level trends
  * Fine skills (Phase 5) → user-facing recommendations

### Data source summary
- postings.csv: main job records
- jobs/job_skills.csv: job_id → skill_abr (category code)
- mappings/skills.csv: skill_abr → skill category name
- skills_desc column is 98% empty (2,439 / 123,849) — DO NOT use for skill extraction

In [11]:
import pandas as pd

job_skills = pd.read_csv("../data/raw/archive/jobs/job_skills.csv")
skills = pd.read_csv("../data/raw/archive/mappings/skills.csv")

print("job_skills:", job_skills.shape)
print("skills:", skills.shape)

job_skills: (213768, 2)
skills: (35, 2)


In [12]:
job_skills.head()

,job_id,skill_abr
0,3884428798,MRKT
1,3884428798,PR
2,3884428798,WRT
3,3887473071,SALE
4,3887465684,FIN


In [13]:
skills.head()

,skill_abr,skill_name
0,ART,Art/Creative
1,DSGN,Design
2,ADVR,Advertising
3,PRDM,Product Management
4,DIST,Distribution


In [14]:
job_skills.info()
skills.info()

<class 'pandas.DataFrame'>
RangeIndex: 213768 entries, 0 to 213767
Data columns (total 2 columns):
 #   Column     Non-Null Count   Dtype
---  ------     --------------   -----
 0   job_id     213768 non-null  int64
 1   skill_abr  213768 non-null  str  
dtypes: int64(1), str(1)
memory usage: 3.3 MB
<class 'pandas.DataFrame'>
RangeIndex: 35 entries, 0 to 34
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   skill_abr   35 non-null     str  
 1   skill_name  35 non-null     str  
dtypes: str(2)
memory usage: 692.0 bytes


In [15]:

job_skills_named = job_skills.merge(skills, on="skill_abr", how="left")

print("job_skills_named:", job_skills_named.shape)
job_skills_named.head()

job_skills_named: (213768, 3)


,job_id,skill_abr,skill_name
0,3884428798,MRKT,Marketing
1,3884428798,PR,Public Relations
2,3884428798,WRT,Writing/Editing
3,3887473071,SALE,Sales
4,3887465684,FIN,Finance


In [16]:

postings_with_skills = postings.merge(job_skills_named, on="job_id", how="left")

print("postings_with_skills:", postings_with_skills.shape)
postings_with_skills.head(20)

postings_with_skills: (207531, 33)


,job_id,company_name,title,description,max_salary,pay_period,location,company_id,views,med_salary,...,posting_domain,sponsored,work_type,currency,compensation_type,normalized_salary,zip_code,fips,skill_abr,skill_name
0,921716,Corcoran Sawyer Smith,Marketing Coordinator,Job descriptionA leading real estate firm in N...,20.0,HOURLY,"Princeton, NJ",2774458.0,20.0,NaN,...,NaN,0,FULL_TIME,USD,BASE_SALARY,38480.0,8540.0,34021.0,MRKT,Marketing
1,921716,Corcoran Sawyer Smith,Marketing Coordinator,Job descriptionA leading real estate firm in N...,20.0,HOURLY,"Princeton, NJ",2774458.0,20.0,NaN,...,NaN,0,FULL_TIME,USD,BASE_SALARY,38480.0,8540.0,34021.0,SALE,Sales
2,1829192,NaN,Mental Health Therapist/Counselor,"At Aspen Therapy and Wellness , we are committ...",50.0,HOURLY,"Fort Collins, CO",NaN,1.0,NaN,...,NaN,0,FULL_TIME,USD,BASE_SALARY,83200.0,80521.0,8069.0,HCPR,Health Care Provider
3,10998357,The National Exemplar,Assitant Restaurant Manager,The National Exemplar is accepting application...,65000.0,YEARLY,"Cincinnati, OH",64896719.0,8.0,NaN,...,NaN,0,FULL_TIME,USD,BASE_SALARY,55000.0,45202.0,39061.0,MGMT,Management
4,10998357,The National Exemplar,Assitant Restaurant Manager,The National Exemplar is accepting application...,65000.0,YEARLY,"Cincinnati, OH",64896719.0,8.0,NaN,...,NaN,0,FULL_TIME,USD,BASE_SALARY,55000.0,45202.0,39061.0,MNFC,Manufacturing
5,23221523,"Abrams Fensterman, LLP",Senior Elder Law / Trusts and Estates Associat...,Senior Associate Attorney - Elder Law / Trusts...,175000.0,YEARLY,"New Hyde Park, NY",766262.0,16.0,NaN,...,NaN,0,FULL_TIME,USD,BASE_SALARY,157500.0,11040.0,36059.0,OTHR,Other
6,35982263,NaN,Service Technician,Looking for HVAC service tech with experience ...,80000.0,YEARLY,"Burlington, IA",NaN,3.0,NaN,...,NaN,0,FULL_TIME,USD,BASE_SALARY,70000.0,52601.0,19057.0,IT,Information Technology
7,91700727,Downtown Raleigh Alliance,Economic Development and Planning Intern,Job summary:The Economic Development & Plannin...,20.0,HOURLY,"Raleigh, NC",1481176.0,9.0,NaN,...,NaN,0,INTERNSHIP,USD,BASE_SALARY,35360.0,27601.0,37183.0,PRJM,Project Management
8,103254301,Raw Cereal,Producer,Company DescriptionRaw Cereal is a creative de...,300000.0,YEARLY,United States,81942316.0,7.0,NaN,...,NaN,0,CONTRACT,USD,BASE_SALARY,180000.0,NaN,NaN,DSGN,Design
9,103254301,Raw Cereal,Producer,Company DescriptionRaw Cereal is a creative de...,300000.0,YEARLY,United States,81942316.0,7.0,NaN,...,NaN,0,CONTRACT,USD,BASE_SALARY,180000.0,NaN,NaN,ART,Art/Creative


In [17]:
postings_with_skills[["job_id", "title", "skill_name"]].head(20)

,job_id,title,skill_name
0,921716,Marketing Coordinator,Marketing
1,921716,Marketing Coordinator,Sales
2,1829192,Mental Health Therapist/Counselor,Health Care Provider
3,10998357,Assitant Restaurant Manager,Management
4,10998357,Assitant Restaurant Manager,Manufacturing
5,23221523,Senior Elder Law / Trusts and Estates Associat...,Other
6,35982263,Service Technician,Information Technology
7,91700727,Economic Development and Planning Intern,Project Management
8,103254301,Producer,Design
9,103254301,Producer,Art/Creative


In [18]:
postings_with_skills["skill_name"].nunique()

35

In [19]:
postings_with_skills["skill_name"].value_counts().head(20)

skill_name
Information Technology    25256
Sales                     21193
Management                20385
Manufacturing             17728
Health Care Provider      16675
Business Development      13304
Engineering               12530
Other                     12314
Finance                    8011
Marketing                  5400
Accounting/Auditing        4999
Administrative             4787
Customer Service           4167
Project Management         3825
Analyst                    3775
Research                   2928
Human Resources            2608
Consulting                 2241
Legal                      2229
Education                  2217
Name: count, dtype: int64